File: internal/segment/segment_test.go
package segment

import (
	"context"
	"testing"

	"github.com/tidepool/tidepool/internal/document"
	"github.com/tidepool/tidepool/internal/storage"
)

func TestSegmentRoundTrip(t *testing.T) {
	store := storage.NewInMemoryStore()
	writer := NewWriter(store, "test")
	reader := NewReader(store, "test", "")

	docs := []*document.Document{
		{ID: "a", Vector: []float32{1, 2}, Attributes: map[string]interface{}{"tag": "x"}},
		{ID: "b", Vector: []float32{3, 4}, Attributes: map[string]interface{}{"tag": "y"}},
	}

	seg, err := writer.WriteSegment(context.Background(), docs)
	if err != nil {
		t.Fatalf("write segment: %v", err)
	}

	loaded, err := reader.ReadSegment(context.Background(), seg.SegmentKey)
	if err != nil {
		t.Fatalf("read segment: %v", err)
	}

	if loaded.Dimensions != 2 {
		t.Fatalf("expected dimensions 2, got %d", loaded.Dimensions)
	}
	if len(loaded.IDs) != 2 {
		t.Fatalf("expected 2 ids, got %d", len(loaded.IDs))
	}
	if loaded.IDs[0] != "a" || loaded.IDs[1] != "b" {
		t.Fatalf("unexpected ids: %v", loaded.IDs)
	}
	if loaded.Vectors[0][0] != 1 || loaded.Vectors[1][1] != 4 {
		t.Fatalf("unexpected vectors: %v", loaded.Vectors)
	}
}

func TestSegmentDimensionMismatch(t *testing.T) {
	store := storage.NewInMemoryStore()
	writer := NewWriter(store, "test")

	docs := []*document.Document{
		{ID: "a", Vector: []float32{1, 2}},
		{ID: "b", Vector: []float32{3}},
	}

	if _, err := writer.WriteSegment(context.Background(), docs); err == nil {
		t.Fatal("expected dimension mismatch error")
	}
}

File: crates/tidepool-common/tests/segment_test.rs
use std::collections::BTreeMap;

use tidepool_common::attributes::AttrValue;
use tidepool_common::document::Document;
use tidepool_common::segment::{Reader, Writer};
use tidepool_common::storage::InMemoryStore;

fn tag_attr(value: &str) -> AttrValue {
    let mut map = BTreeMap::new();
    map.insert("tag".to_string(), AttrValue::String(value.to_string()));
    AttrValue::Object(map)
}

#[tokio::test]
async fn segment_roundtrip() {
    let store = InMemoryStore::new();
    let writer = Writer::new(store.clone(), "test");
    let reader = Reader::new(store.clone(), "test", None);

    let docs = vec![
        Document {
            id: "a".to_string(),
            vector: vec![1.0, 2.0],
            attributes: Some(tag_attr("x")),
        },
        Document {
            id: "b".to_string(),
            vector: vec![3.0, 4.0],
            attributes: Some(tag_attr("y")),
        },
    ];

    let seg = writer.write_segment(&docs).await.unwrap().unwrap();
    let loaded = reader.read_segment(&seg.segment_key).await.unwrap();

    assert_eq!(loaded.dimensions, 2);
    assert_eq!(loaded.ids.len(), 2);
    assert_eq!(loaded.ids[0], "a");
    assert_eq!(loaded.ids[1], "b");
    assert_eq!(loaded.vectors[0][0], 1.0);
    assert_eq!(loaded.vectors[1][1], 4.0);
    assert!(loaded.index.is_some());
}

#[tokio::test]
async fn segment_dimension_mismatch() {
    let store = InMemoryStore::new();
    let writer = Writer::new(store, "test");

    let docs = vec![
        Document {
            id: "a".to_string(),
            vector: vec![1.0, 2.0],
            attributes: None,
        },
        Document {
            id: "b".to_string(),
            vector: vec![3.0],
            attributes: None,
        },
    ];

    let result = writer.write_segment(&docs).await;
    assert!(result.is_err());
}
